### PR TITLE
mgr/dashboard_v2: Import/add initial CephPool component

### PR DESCRIPTION
**Pull request by s0nea**
This pull request is intended to add a 'view only' representation of the CephPool object. That means all operations e.g. create, delete and so on are missing here and will be added later.

* Import CephPool model
* Add CephPool controller

CephPool tests are missing because we don't have any 'view only' testcase currently.

Signed-off-by: Tatjana Dehler <tdehler@suse.com>

### CODE REVIEW

File: src/pybind/mgr/dashboard_v2/controllers/pool.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+from functools import partial
+
+from ..models.nodb import nodb_context
+from ..models.pool import CephPool
+from ..tools import ApiController, RESTController
+
+
+def nodb_serializer(model, obj):
```
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard_v2/controllers/pool.py]: Can you move `nodb_serializer` to `..models.nodb` and just import it here and in `.cluster`?

**Comment by sebastian-philipp** [comment]
> CephPool tests are missing because we don't have any 'view only' testcase currently.

I think it makes sense to also include our other Pool tests that are independent of the RESTController.

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard_v2/controllers/pool.py (hunk 1 above)
Comment (s0nea) [reply src/pybind/mgr/dashboard_v2/controllers/pool.py]: Yes, sure

**Comment by jecluis** [comment]
@s0nea would you mind adding more info to the commit body, telling us what you are doing and why this commit is important? It is a good habit considering the commit's body as a small piece of insight into what the commit is achieving, and why it is needed.

It may not be beneficial to those from the oA team, who may very well know what you are doing, but those in Ceph who will be reviewing these commits may have a lot to benefit from.

And although I'm mentioning this here, now, I think this also applies to (possibly almost) every other pull request from other team members :)

**Comment by s0nea** [comment]
@jecluis no, I don't mind and I agree. I will add some more description to the commit message.

**Comment by jecluis** [comment]
@s0nea thanks.

Also, why are we grabbing osd/pool information from the monitors when we have that information through the in-memory osdmap, directly from the manager?

**Comment by s0nea** [comment]
@jecluis because that's how it was handled by openATTIC before and we agreed on migrating the models as they are (at least from my understanding). I'm not really sure if we want to/should improve these things now or in a second step?

**Comment by jecluis** [comment]
@s0nea this just seems like a lot of clutter to achieve, basically, the same thing as a call to MgrModule.get(). Except the latter is accessing in-memory, already available information, and what we're doing here is reaching out to the monitors without a clear added benefit. Can someone please expand on why this is preferable over using the in-memory structures, please?

Also, keep in mind that this will have to be reviewed by the upstream team; if they are already a bit off-put by the idea of having a proper rest api, instead preferring to directly access the in-memory data, can you all imagine what their reaction will be once you're reaching out to the monitors? :)

Maybe @sebastian-philipp will want to chime in here.

**Comment by s0nea** [comment]
@jecluis with 'a second step' I didn't mean it won't ever happen ;) It's just the groundwork, so someone can work on the frontend part for example, and in the meanwhile all these things can be improved.

**Comment by jecluis** [comment]
@s0nea if this is going to be "improved", why can't we perform said improvements now instead of waiting for *whenever*? I mean, what's holding you from calling the MgrModule.get() function now, much like what the current dashboard does (minus the abstractions they create), and populate whatever data structure you want from that?

File: src/pybind/mgr/dashboard_v2/models/pool.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+                     else None),
+                'last_change': pool_data['last_change'],
+                'full': 'full' in pool_data['flags_names'],
+                'min_size': pool_data['min_size'],
+                # TODO: Check what happened to this value.
+                # It leads to 'KeyError: 'crash_replay_interval' currently
+                # 'crash_replay_interval': pool_data['crash_replay_interval'],
+                'pg_num': pool_data['pg_num'],
+                'size': pool_data['size'],
+                'crush_ruleset': pool_data[
```
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard_v2/models/pool.py]: Finally, we can unconditionally rename `crush_ruleset` to `crush_rule`.

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]

**Comment by sebastian-philipp** [comment]
@jecluis I'm still favoring the idea of copying as much as possible from the existing code base, providing an API for that and use the existing UI. As soon as we migrate the edit dialog into dashboard_v2, we can just copy the old save() method. 

That doesn't mean we can't improve our code, though. I had to modify our [CephCluster model in a similar way](https://github.com/openattic/ceph/blob/wip-mgr-dashboard_v2/src/pybind/mgr/dashboard_v2/models/cluster.py#L55). @jecluis If you expect the PR will raise concerns, we should fix it, of course.

**Comment by jecluis** [comment]
@sebastian-philipp so, correct me if I'm wrong, but this whole thing seems a bit over-engineering, and it also feels like this is one of those cases that just doing it right the first time would have actually saved time. I mean, look at the `CephPool` class:

1. it's basically declaring variables for every single field in `pg_pool_t`
2. it's making a monitor call to fetch an osd dump
3. it's then initiating each variable using the result of said osd dump

How is this easier than calling a `get()` for the osdmap, and simply save the map, or whatever it is returned? Alas, if the return is in the form of an object, you don't even have to initiate any variables, as your variable names are already mapping to what the dump returns. If it's not an object, I'd even argue providing the info directly from the map would be cleaner.

Even if the plan is to "improve" this down the line, when is said improvement going to happen? Before you open the initial PR, I hope. But look at all that duplicated work; what's the benefit in that? And what's the cost of doing things in the "improved" way on the first pass?

**Review by votdev — DISMISSED** [review]

**Review by jecluis — CHANGES_REQUESTED** [review]
We have two logical changes in this PR, which belong in two different commits.

I suggest the following division (first commit to last commit):

1. `Move nodb_serializer into nodb module` should only contain the removal of `nodb_serializer()` from `cluster.py` and the addition of said function into `nodb.py`.
2. `Import/add initial CephPool component`, containing solely the changes to the `pool.py` model and controller.

**Comment by jecluis** [comment]
Also, @s0nea, are there any tests exercising the code you're adding on this patchset? If not, mind adding them?

File: src/pybind/mgr/dashboard_v2/models/pool.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+                'erasure_code_profile_id':
+                    (pool_data['erasure_code_profile'] if 'erasure_code_profile' in pool_data
+                     else None),
+                'last_change': pool_data['last_change'],
+                'full': 'full' in pool_data['flags_names'],
+                'min_size': pool_data['min_size'],
+                # TODO: 'crash_replay_interval': pool_data['crash_replay_interval'],
+                'pg_num': pool_data['pg_num'],
+                'size': pool_data['size'],
+                'crush_ruleset': pool_data['crush_rule'],
```
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard_v2/models/pool.py]: We should rename this to `crush_rule`.

File: src/pybind/mgr/dashboard_v2/models/pool.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+                # Considered advanced options
+                'pgp_num': pool_data['pg_placement_num'],
+                'stripe_width': pool_data['stripe_width'],
+                'quota_max_bytes': pool_data['quota_max_bytes'],
+                'quota_max_objects': pool_data['quota_max_objects'],
+                # Cache tiering related
+                'cache_mode': pool_data['cache_mode'],
+                'tier_of_id': pool_data['tier_of'] if pool_data['tier_of'] > 0 else None,
```
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard_v2/models/pool.py]: `s/tier_of_id/tier_of/`. The `_id` suffix doesn't work right now, as we now have IntegerFields. Dito for `write_tier_id` and `read_tier_id`

File: src/pybind/mgr/dashboard_v2/models/pool.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+                'id': pool_data['pool'],
+                'name': pool_data['pool_name'],
+                'type': {1: 'replicated', 3: 'erasure'}[pool_data['type']],  # type is an
+                                                                             # undocumented dump of
+                # https://github.com/ceph/ceph/blob/289c10c9c79c46f7a29b5d2135e3e4302ac378b0/src/osd/osd_types.h#L1035
+                'erasure_code_profile_id':
+                    (pool_data['erasure_code_profile'] if 'erasure_code_profile' in pool_data
+                     else None),
+                'last_change': pool_data['last_change'],
+                'full': 'full' in pool_data['flags_names'],
```
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard_v2/models/pool.py]: We export `flags_names` already in line 112. I'd remove `full` here.

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard_v2/models/pool.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+            if pool.id in df_per_pool:
+                pool.max_avail = df_per_pool[pool.id]['max_avail']
+                pool.kb_used = df_per_pool[pool.id]['kb_used']
+                pool.percent_used = df_per_pool[pool.id]['percent_used']
+            else:
+                pool.max_avail = None
+                pool.kb_used = None
+                pool.percent_used = None
+
+    def __unicode__(self):
```
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard_v2/models/pool.py]: for Python 3, we also need `__str__`

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard_v2/models/pool.py (hunk 3 above)
Comment (s0nea) [reply src/pybind/mgr/dashboard_v2/models/pool.py]: Ah, okay, so we do also rename the API fields? - Good to know :)

File: src/pybind/mgr/dashboard_v2/models/pool.py (hunk 5 above)
Comment (s0nea) [reply src/pybind/mgr/dashboard_v2/models/pool.py]: Jap, agree.

File: src/pybind/mgr/dashboard_v2/models/pool.py (hunk 6 above)
Comment (s0nea) [reply src/pybind/mgr/dashboard_v2/models/pool.py]: Okay, thanks.

File: src/pybind/mgr/dashboard_v2/models/pool.py (hunk 4 above)
Comment (s0nea) [reply src/pybind/mgr/dashboard_v2/models/pool.py]: Okay, I will adapt it.

**Review by votdev — DISMISSED** [review]

**Review by p-se — DISMISSED** [review]

**Comment by jecluis** [comment]
As per our conversation on IRC, @s0nea agrees this PR should be held from being merged until after we open the initial PR to ceph/master

**Comment by s0nea** [comment]
I'm closing this pull request because we are not migrating the NoDB models to dashboard_v2 currently. If the code becomes relevant again, I will open this pull request once more.

**Pull request closed without merging**